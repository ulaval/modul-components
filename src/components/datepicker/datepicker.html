<div class="m-datepicker"
     :class="{ 'm--is-disabled': isDisabled,
               'm--is-waiting': isWaiting,
               'm--has-error': calandarError,
               'm--is-valid': isValid }">
    <m-input-style :label="label"
                   :disabled="isDisabled"
                   :waiting="isWaiting"
                   :focus="open"
                   :empty="isEmpty"
                   :error="calandarError"
                   :valid="isValid"
                   :icon-name="iconName"
                   v-m-popup:popup>
        <input type="text"
               ref="input"
               :disabled="!active"
               :placeholder="placeholder"
               @keydown.enter="inputOnKeydownEnter"
               @keydown.return="inputOnKeydownEnter"
               @keydown.esc="inputOnKeydownEsc"
               @keydown.tab="inputOnKeydownTab"
               @keydown="inputOnKeydown"
               @focus="inputOnFocus"
               @change="validateDate"
               v-model="formattedDate">
        <m-icon-button slot="right-content"
                       icon-name="calendar"
                       :disabled="!active"
                       button-size="22px"
                       @keydown.enter.prevent="open = !isDisabled">
            <m-i18n k="m-datepicker:open" v-if="!open"></m-i18n>
            <m-i18n k="m-datepicker:close" v-if="open"></m-i18n>
        </m-icon-button>
    </m-input-style>
    <m-validation-message
        :disabled="isDisabled"
        :waiting="isWaiting"
        :error="calandarError"
        :error-message="calandarErrorMessage"
        :valid-message="validMessage"
        :helper-message="helperMessage"></m-validation-message>
    <m-popup ref="popup"
             :open.sync="open"
             :disabled="!active"
             :close-on-backdrop="true"
             :focus-management="isMqMaxS"
             :padding="false"
             placement="bottom">
        <div slot="header" class="m-datepicker__header">
            <nav class="m-datepicker__header__nav">
                <m-icon-button icon-name="left-chevron" @click="selectYear(selectedYear - 1)" :disabled="isMinYear">previous year</m-icon-button>
                <button type="button" class="m-datepicker__years" @click="showYears">{{selectedYear}}</button>
                <m-icon-button icon-name="chevron" @click="selectYear(selectedYear + 1)" :disabled="isMaxYear">next year</m-icon-button>
            </nav>
            <nav class="m-datepicker__header__nav">
                <m-icon-button icon-name="left-chevron" @click="selectMonth(selectedMonth - 1)" :disabled="isMinMonth">previous month</m-icon-button>
                <button type="button" class="m-datepicker__months" @click="view = 'month'">{{ selectedMonthName }}</button>
                <m-icon-button icon-name="chevron" @click="selectMonth(selectedMonth + 1)" :disabled="isMaxMonth">next month</m-icon-button>
            </nav>
        </div>
        <div class="m-datepicker__body" ref="body">
            <table class="m-datepicker__body-table">
                <template v-if="view === 'year'">
                    <tr v-for="yearRow in years">
                        <td v-for="y in yearRow" class="m-datepicker__body-cell" v-if="y !== undefined">
                            <button type="button"
                                    class="m-datepicker__item"
                                    :class="{'m--is-selected': selectedYear === y}"
                                    @click.stop="selectYear(y, true)">
                                <span class="m-datepicker__item__text">{{ y }}</span>
                            </button>
                        </td>
                    </tr>
                </template>
                <template v-if="view === 'month'">
                    <tr v-for="monthRow in months">
                        <td v-for="m in monthRow" class="m-datepicker__body-cell">
                            <button type="button"
                                    class="m-datepicker__item"
                                    :class="{'m--is-selected': selectedMonthName === m.name,
                                                'm--is-disabled': m.isDisabled}"
                                    :disabled="m.isDisabled"
                                    @click.stop="selectMonth(m.index, true)">
                                <span class="m-datepicker__item__text">{{ m.name }}</span>
                            </button>
                        </td>
                    </tr>
                </template>
                <template v-if="view === 'day'">
                    <tr>
                        <td v-for="w in weekdays" class="m-datepicker__weekday">{{ w }}</td>
                    </tr>
                    <tr v-for="dayRow in daysOfMonth">
                        <td v-for="d in dayRow" class="m-datepicker__body-cell">
                            <button type="button"
                                    class="m-datepicker__item"
                                    :class="{ 'm--is-selected': d.isSelected,
                                                'm--is-today': d.isToday,
                                                'm--is-disabled': d.isDisabled,
                                                'm--is-other-month': d.month !== selectedMonth }"
                                    @click.stop="selectDate(d)">
                                <span class="m-datepicker__item__text">{{ d.date }}</span>
                            </button>
                        </td>
                    </tr>
                </template>
            </table>
        </div>
    </m-popup>
</div>
