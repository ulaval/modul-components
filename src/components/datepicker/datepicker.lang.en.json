{
    "m-datepicker": {
        "placeholder": "YYYY-MM-DD",
        "open": "Open calendar",
        "close": "Close calendar",
        "out-of-range-error": "Date is out of range",
        "format-error": "Invalid format"
    }
}
