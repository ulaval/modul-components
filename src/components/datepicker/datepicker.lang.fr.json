{
    "m-datepicker": {
        "placeholder-aaaa-mm-jj": "AAAA-MM-JJ",
        "placeholder-aaaa-mm": "AAAA-MM",
        "open": "Ouvrir le calendrier",
        "close": "Fermer le calendrier",
        "out-of-range-error": "Date hors limites",
        "format-error": "Entrez une date au format AAAA-MM-JJ"
    }
}
