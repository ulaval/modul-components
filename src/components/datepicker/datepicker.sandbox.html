<div>
    <h2>Default datepicker</h2>
    <m-datepicker class="m-u--margint-top"></m-datepicker>

    <h2>Add error message on blur event</h2>
    <m-datepicker class="m-u--margint-top"
                  @blur="addMessageError"
                  @open="onOpen"
                  @close="onClose"
                  :error-message="errorMessage">
    </m-datepicker>

    <h2>Disabled</h2>
    <m-datepicker class="m-u--margint-top"
                  :disabled="true">
    </m-datepicker>

    <h2>Selected value</h2>
    <m-datepicker class="m-u--margint-top"
                  v-model="model"></m-datepicker>
    <p>Selected value: {{ model }}</p>
</div>
