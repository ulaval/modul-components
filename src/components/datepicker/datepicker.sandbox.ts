import Vue, { PluginObject } from 'vue';
import { Component } from 'vue-property-decorator';

import { DATEPICKER_NAME } from '../component-names';
import WithRender from './datepicker.sandbox.html';

@WithRender
@Component
export class MDatepickerSandbox extends Vue {
}

const DatepickerSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${DATEPICKER_NAME}-sandbox`, MDatepickerSandbox);
    }
};

export default DatepickerSandboxPlugin;
