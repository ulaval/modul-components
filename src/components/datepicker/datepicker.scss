@import '../../styles/abstracts/variables';
$m-datepicker--item-dimension: $m-touch-size !default;

.m-datepicker {
    display: inline-flex;
    flex-direction: column;

    &__header,
    &__body {
        background: $m-color--white;

        @media (max-width: $m-mq--max--s) {
            width: 100%;
        }

        @media (min-width: $m-mq--min--s) {
            width: 7 * $m-datepicker--item-dimension + 2 * $m-padding--s + 2;
        }
    }

    &__header {
        justify-content: space-around;
        padding: $m-padding--s 0;

        @media (max-width: $m-mq--max--s) {
            border-bottom: $m-border-width--s solid $m-color--border;
        }

        @media (min-width: $m-mq--min--s) {
            border: $m-border-width--s solid $m-color--border;
        }

        &,
        &__nav {
            display: flex;
            align-items: center;
        }

        &__nav {
            justify-content: space-between;
            width: 50%;
        }
    }

    &__years,
    &__months,
    &__item {
        background: none;
        border: 0;
        color: inherit;
        font: inherit;
        outline: none;
        padding: 0;
        cursor: pointer;
    }

    &__years,
    &__months {
        transition: color $m-transition-duration ease;
        font-weight: $m-font-weight--bold;
        color: $m-color--text;

        &:hover,
        &:focus {
            color: $m-color--grey-darker;
        }
    }

    &__body {
        display: flex;
        justify-content: center;
        overflow-y: auto;
        height: $m-datepicker--item-dimension * 7 + $m-padding + $m-border-width--s;
        padding: $m-padding--s;

        @media (min-width: $m-mq--min--s) {
            border-right: $m-border-width--s solid $m-color--border;
            border-bottom: $m-border-width--s solid $m-color--border;
            border-left: $m-border-width--s solid $m-color--border;
        }

        &-table {
            border-spacing: 0;
            border-collapse: collapse;
            text-align: center;
            width: 100%;
        }

        &-cell {
            padding: 0;
        }
    }

    &__weekday {
        padding: 0;
        width: $m-datepicker--item-dimension;
        height: $m-datepicker--item-dimension;
        font-weight: $m-font-weight--bold;
    }

    &__item {
        position: relative;
        width: $m-datepicker--item-dimension;
        height: $m-datepicker--item-dimension;

        &__text {
            position: relative;
            z-index: 1;
        }

        &::after {
            transition: transform $m-transition-duration ease, opacity $m-transition-duration ease;
            position: absolute;
            top: -50%;
            left: -50%;
            z-index: 0;
            transform: translate3d(50%, 50%, 0) scale(0);
            content: '';
            border-radius: 50%;
            width: $m-touch-size;
            height: $m-touch-size;
            background: $m-color--interactive-lightest;
            opacity: 0;
        }

        &.m--is-today:not(.m--is-selected) {
            &::after {
                background: $m-color--grey-light;
                font-weight: $m-font-weight--bold;
            }
        }

        &:hover,
        &:focus,
        &.m--is-selected,
        &.m--is-today {
            &::after {
                transform: translate3d(50%, 50%, 0) scale(1);
                opacity: 1;
            }
        }

        &.m--is-selected {
            color: white;
            font-weight: $m-font-weight--bold;

            &::after {
                background: $m-color--interactive-light;
            }
        }

        &.m--is-other-month {
            transition: color $m-transition-duration ease;
            color: $m-color--disabled;

            &:hover,
            &:focus {
                color: $m-color--text;
            }
        }
    }

    .m--is-disabled {
        background-color: none;
        color: $m-color--grey-light;
        cursor: not-allowed;
    }
}
