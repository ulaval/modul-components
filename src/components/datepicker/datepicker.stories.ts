import { withA11y } from '@storybook/addon-a11y';
import { withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/vue';
import Vue from 'vue';
import { componentsHierarchyRootSeparator } from '../../../conf/storybook/utils';
import { DATEPICKER_NAME } from '../component-names';
import DatepickerPlugin from './datepicker';

Vue.use(DatepickerPlugin);

declare module '@storybook/addon-knobs' {
    export function withKnobs(): any;
}

storiesOf(`${componentsHierarchyRootSeparator}${DATEPICKER_NAME}`, module)
    .addDecorator(withA11y)
    .addDecorator(withKnobs)
    .add('default', () => ({
        template: `<m-datepicker></m-datepicker>`
    }))
    .add('events', () => ({
        data: () => ({
            model1: '2011-01-01'
        }),
        methods: {
            onInputChange(value: string): string {
                // tslint:disable-next-line: no-console
                console.log('MDatePicker.onInputChange=' + value);
                return value;
            },
            onFocus(value: Event): void {
                // tslint:disable-next-line: no-console
                console.log('MDatePicker.onFocus!');

            },
            onBlur(event: Event): void {
                // tslint:disable-next-line: no-console
                console.log('MDatePicker.onBlur!');

            }
        },
        template: `<div><m-datepicker :value="model1" @change="model1 = onInputChange($event)" @focus="onFocus" @blur="onBlur"></m-datepicker> <br/><br/>model value = {{model1}}</div>`
    }))
    .add('label', () => ({
        template: `<m-datepicker label="Date label"></m-datepicker>`
    }))

    .add('placeholder', () => ({
        template: `<m-datepicker placeholder="Lorem Ipsum"></m-datepicker>`
    }))

    .add('waiting', () => ({
        template: `<m-datepicker :waiting="true"></m-datepicker>`
    }))

    .add('min="2008-01-01" && max="2014-12-31"', () => ({
        data: () => ({
            model1: '2011-01-01'
        }),
        template: `<div><m-datepicker min="2008-01-01" max="2014-12-31" v-model="model1"></m-datepicker>model value = {{model1}}</div>`
    }))
    .add('date format invalid', () => ({
        data: () => ({
            model1: '2000-19-12'
        }),
        template: `<div><m-datepicker min="2008-01-01" max="2014-12-31" v-model="model1"></m-datepicker>model value = {{model1}}</div>`
    }))
    .add('date off limit min', () => ({
        data: () => ({
            model1: '2000-01-01'
        }),
        template: `<div><m-datepicker min="2008-01-01" max="2014-12-31" v-model="model1"></m-datepicker>model value = {{model1}}</div>`
    }))
    .add('date off limit  max', () => ({
        data: () => ({
            model1: '2015-01-01'
        }),
        template: `<div><m-datepicker min="2008-01-01" max="2014-12-31" v-model="model1"></m-datepicker>model value = {{model1}}</div>`
    }))

    .add('disabled', () => ({
        template: `<m-datepicker :disabled="true"></m-datepicker>`
    }))
    .add('readonly', () => ({
        template: `<m-datepicker :readonly="true"></m-datepicker>`
    }))
    .add('valid', () => ({
        template: `<m-datepicker :valid="true"></m-datepicker>`
    }))
    .add('error-message', () => ({
        template: `<m-datepicker :error="true" error-message="this is an error"></m-datepicker>`
    }))
    .add('helper-message', () => ({
        template: `<m-datepicker helper-message="AAAA-MM-JJ"></m-datepicker>`
    }))
    .add('hide-internal-error-message', () => ({
        data: () => ({
            model1: '9999-99-99'
        }),
        template: `<m-datepicker v-model="model1" :hide-internal-error-message="true"></m-datepicker>`
    }));

