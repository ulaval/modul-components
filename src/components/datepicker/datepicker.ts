import { ModulVue } from '../../utils/vue/vue';
import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop, Model } from 'vue-property-decorator';
import WithRender from './datepicker.html?style=./datepicker.scss';
import { DATEPICKER_NAME } from '../component-names';
import * as moment from 'moment';
import i18nPlugin, { curLang } from '../../utils/i18n/i18n';
import { InputState } from '../../mixins/input-state/input-state';
import { InputPopup } from '../../mixins/input-popup/input-popup';
import { MediaQueries } from '../../mixins/media-queries/media-queries';
import { KeyCode } from '../../utils/keycode/keycode';
import MediaQueriesPlugin from '../../utils/media-queries/media-queries';
import ButtonPlugin from '../button/button';
import InputStylePlugin, { MInputStyle } from '../input-style/input-style';
import ValidationMessagePlugin from '../validation-message/validation-message';
import PopupPlugin from '../popup/popup';

const VIEW_DAY = 'day';
const VIEW_MONTH = 'month';
const VIEW_YEAR = 'year';
const NB_YEARS_PER_ROW = 5;
const ITEM_DIMENSION = 40;

export interface DatepickerDate {
    date: number;
    month: number;
    year: number;
    isDisabled: boolean;
    isToday: boolean;
    isSelected: boolean;
}

@WithRender
@Component({
    mixins: [
        InputState,
        InputPopup,
        MediaQueries
    ]
})
export class MDatepicker extends ModulVue {

    @Model('change')
    @Prop()
    public value: moment.Moment | Date;
    @Prop()
    public label: string;
    @Prop()
    public iconName: string;
    @Prop()
    public required: boolean;
    @Prop({ default: 'YYYY/MM/DD' })
    public format: string;
    @Prop({ default: () => { return moment().subtract(10, 'year'); } })
    public min: moment.Moment | Date;
    @Prop({ default: () => { return moment().add(10, 'year'); } })
    public max: moment.Moment | Date;

    public placeholder: string = this.$i18n.translate('m-datepicker:placeholder');
    private internalOpen: boolean = false;
    private view: string = 'day';
    private previousDays: DatepickerDate[] = [];
    private days: DatepickerDate[] = [];
    private nextDays: DatepickerDate[] = [];
    private selectedYear: number = 0;
    private selectedMonth: number = 0;
    private selectedDay: number = 0;

    private mouseIsDown: boolean = false;
    private internalCalandarErrorMessage: string = '';

    protected created(): void {
        moment.locale([curLang, 'en-ca']);
        this.selectedMomentDate = this.valueIsValid() ? moment(this.value) : moment();
    }

    private valueIsValid(): boolean {
        return this.value && moment(this.value).isValid() && moment(this.value).isBetween(this.min, this.max, 'day', '[]');
    }

    private get years(): number[] {
        let years: number[] = [];
        for (let year = moment(this.max).year(); year >= moment(this.min).year(); year--) {
            years.push(year);
        }
        return this.prepareDataForTableLayout(years, NB_YEARS_PER_ROW);
    }

    private get months(): any[] {
        let months: any[] = [];
        for (let index = 0; index <= 11; index++) {
            months.push({
                index,
                name: moment.monthsShort()[index],
                isDisabled: !moment({ year: this.selectedMomentDate.year(), month: index }).isBetween(this.min, this.max, 'month', '[]')
            });
        }
        return this.prepareDataForTableLayout(months, 3);
    }

    private get weekdays(): string[] {
        return moment.weekdaysMin();
    }

    private getDaysOfPreviousMonth(): DatepickerDate[] {
        let monthStartsAt = moment(this.selectedMomentDate).startOf('month').weekday();
        let days: DatepickerDate[] = [];

        for (let index = monthStartsAt; index > 0; index--) {
            let date = moment(this.selectedMomentDate).startOf('month').subtract(index, 'days');
            days.push({
                date: date.date(),
                month: date.month(),
                year: date.year(),
                isDisabled: moment(date).isBefore(this.min, 'day') || moment(date).isAfter(this.max, 'day'),
                isToday: false,
                isSelected: false
            });
        }
        return days;
    }

    private getDaysOfCurrentMonth(): DatepickerDate[] {
        let lastDayOfMonth: number = this.selectedMomentDate.daysInMonth();
        let days: DatepickerDate[] = [];

        for (let index = 1; index <= lastDayOfMonth; index++) {
            let date = { year: this.selectedMomentDate.year(), month: this.selectedMomentDate.month(), date: index };
            days.push({
                ...date,
                isDisabled: moment(date).isBefore(this.min, 'day') || moment(date).isAfter(this.max, 'day'),
                isToday: moment().isSame(date, 'day'),
                isSelected: this.selectedMomentDate.isSame(date, 'day')
            });
        }
        return days;
    }

    private getDaysOfNextMonth(): DatepickerDate[] {
        let daysToDisplayFromNextMonth = 6 - moment(this.selectedMomentDate).endOf('month').weekday();
        let days: DatepickerDate[] = [];

        for (let index = 1; index <= daysToDisplayFromNextMonth; index++) {
            let date = moment(this.selectedMomentDate).endOf('month').add(index, 'days');
            days.push({
                date: date.date(),
                month: date.month(),
                year: date.year(),
                isDisabled: moment(date).isBefore(this.min, 'day') || moment(date).isAfter(this.max, 'day'),
                isToday: false,
                isSelected: false
            });
        }
        return days;
    }

    private prepareDataForTableLayout(data: any[], nbItemPerRow: number): any[] {
        let nbRow: number = Math.ceil(data.length / nbItemPerRow);
        let dataTable: any[] = [];
        let count: number = 0;
        for (let row = 0; row < nbRow; row++) {
            let newRow: any[] = [];
            for (let index = 0; index < nbItemPerRow; index++) {
                newRow.push(data[count]);
                count++;
            }
            dataTable.push(newRow);
        }
        return dataTable;
    }

    private get formattedDate(): string {
        return this.internalCalandarErrorMessage || !this.value ? this.as<InputPopup>().internalValue : moment(this.value).format(this.format);
    }

    private set formattedDate(value: string) {
        this.as<InputPopup>().internalValue = value;
    }

    private get selectedMomentDate(): moment.Moment {
        return this.selectedDay ? moment({ year: this.selectedYear, month: this.selectedMonth, day: this.selectedDay }) : moment();
    }

    private set selectedMomentDate(value: moment.Moment) {
        this.selectedYear = value.year();
        this.selectedMonth = value.month();
        this.selectedDay = value.date();
    }

    private get isMinYear(): boolean {
        return this.selectedMomentDate.isSameOrBefore(this.min, 'year');
    }

    private get isMaxYear(): boolean {
        return this.selectedMomentDate.isSameOrAfter(this.max, 'year');
    }

    private get isMinMonth(): boolean {
        return this.selectedMomentDate.isSameOrBefore(this.min, 'month');
    }

    private get isMaxMonth(): boolean {
        return this.selectedMomentDate.isSameOrAfter(this.max, 'month');
    }

    private get selectedMonthName(): string {
        return moment().month(this.selectedMonth).format('MMM');
    }

    private get daysOfMonth(): DatepickerDate[] {
        this.previousDays = this.getDaysOfPreviousMonth();
        this.days = this.getDaysOfCurrentMonth();
        this.nextDays = this.getDaysOfNextMonth();
        return this.prepareDataForTableLayout([...this.previousDays, ...this.days, ...this.nextDays], 7);
    }

    private get calandarError(): boolean {
        return this.internalCalandarErrorMessage != '' || this.as<InputState>().hasError;
    }

    private get calandarErrorMessage(): string {
        return this.as<InputState>().errorMessage != undefined ? this.as<InputState>().errorMessage : this.internalCalandarErrorMessage;
    }

    private get open(): boolean {
        return this.internalOpen;
    }

    private set open(open: boolean) {
        this.internalOpen = open;
        this.$nextTick(() => {
            if (this.internalOpen) {
                let inputEl: any = this.$refs.input;
                inputEl.focus();
                inputEl.setSelectionRange(0, this.formattedDate.length);
                this.$emit('open');
            } else {
                this.$emit('close');
            }
        });
    }

    private validateDate(event): void {
        if (event.target.value == '') {
            this.selectedMomentDate = moment();
            if (this.required) {
                this.internalCalandarErrorMessage = this.$i18n.translate('m-datepicker:required-error');
            } else {
                this.$emit('change', '');
                this.internalCalandarErrorMessage = '';
            }
        } else if (moment(event.target.value, this.format).isValid()) {
            let newDate = moment(event.target.value, this.format);
            if (newDate.isBetween(this.min, this.max, 'day', '[]')) {
                this.selectedMomentDate = newDate;
                this.formattedDate = this.selectedMomentDate.format(this.format);
                this.$emit('change', newDate);
                this.internalCalandarErrorMessage = '';
            } else {
                this.formattedDate = newDate.format(this.format);
                this.internalCalandarErrorMessage = this.$i18n.translate('m-datepicker:out-of-range-error');
            }
        } else {
            this.internalCalandarErrorMessage = this.$i18n.translate('m-datepicker:format-error');
        }
    }

    private keepDateInRange(date: moment.Moment): moment.Moment {
        if (date.isBefore(this.min, 'day')) {
            return moment(this.min);
        } else if (date.isAfter(this.max, 'day')) {
            return moment(this.max);
        }
        return date;
    }

    private showYears(): void {
        this.view = VIEW_YEAR;
        let scrollTop = (Math.floor((moment(this.max).year() - this.selectedYear) / NB_YEARS_PER_ROW)) * ITEM_DIMENSION - (3 * ITEM_DIMENSION);
        setTimeout(() => {
            (this.$refs.body as Element).scrollTo(0, scrollTop);
        }, 10);
    }

    private selectYear(year: number, showMonths: boolean = false): void {
        this.selectedMomentDate = this.keepDateInRange(moment(this.selectedMomentDate).year(year));
        if (showMonths) this.view = VIEW_MONTH;
    }

    private selectMonth(month: number, showDays: boolean = false): void {
        this.selectedMomentDate = this.keepDateInRange(moment(this.selectedMomentDate).month(month));
        if (showDays) this.view = VIEW_DAY;
    }

    private selectDate(selectedDate: DatepickerDate): void {
        if (!selectedDate.isDisabled) {
            this.selectedMomentDate = moment(selectedDate);
            this.internalCalandarErrorMessage = '';
            this.formattedDate = this.selectedMomentDate.format(this.format);
            this.$emit('change', this.value instanceof Date ? this.selectedMomentDate.toDate() : this.selectedMomentDate);
            this.open = false;
        }
    }
}

const DatepickerPlugin: PluginObject<any> = {
    install(v, options): void {
        v.use(InputStylePlugin);
        v.use(ButtonPlugin);
        v.use(PopupPlugin);
        v.use(ValidationMessagePlugin);
        v.use(MediaQueriesPlugin);
        v.use(i18nPlugin);
        v.component(DATEPICKER_NAME, MDatepicker);
    }
};

export default DatepickerPlugin;
