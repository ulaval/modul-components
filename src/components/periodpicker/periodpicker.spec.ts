import { shallowMount, Wrapper } from '@vue/test-utils';
import Vue from 'vue';
import ModulDate from '../../utils/modul-date/modul-date';
import { DatePickerSupportedTypes } from './../datepicker/datepicker';
import PeriodpickerPlugin, { MDateRange, MPeriodpicker, MPeriodpickerFromSlotProps, MPeriodpickerToSlotProps } from './periodpicker';

let wrapper: Wrapper<MPeriodpicker>;
let fromDateProp: DatePickerSupportedTypes;
let toDateProp: DatePickerSupportedTypes;
let minDateProp: DatePickerSupportedTypes;
let maxDateProp: DatePickerSupportedTypes;

let currentFromScopeProps: MPeriodpickerFromSlotProps;
let currentToScopeProps: MPeriodpickerToSlotProps;

const initializeWrapper: () => Wrapper<MPeriodpicker> = () => {
    wrapper = shallowMount(MPeriodpicker, {
        propsData: {
            value: {
                from: fromDateProp,
                to: toDateProp
            },
            min: minDateProp,
            max: maxDateProp
        },
        scopedSlots: {
            first(props: MPeriodpickerFromSlotProps): void {
                currentFromScopeProps = props;
            },
            second(props: MPeriodpickerToSlotProps): void {
                currentToScopeProps = props;
            }
        }
    });

    return wrapper;
};

const expectSelectionEnd: (expectedValue?: MDateRange) => void = (expectedValue: MDateRange) => {
    expect(wrapper.vm.fromIsFocused).toBe(false);
    expect(wrapper.vm.toIsFocused).toBe(false);

    if (expectedValue) {
        expect(wrapper.emitted('input')[0][0]).toEqual(expectedValue);
    }
};

const expectNoSelectionEnd: () => void = () => {
    expect(wrapper.emitted('input')).toBeUndefined();
};

describe(`m-periodpicker`, () => {
    beforeEach(() => {
        Vue.use(PeriodpickerPlugin);

        wrapper = undefined as any;
        fromDateProp = undefined as any;
        toDateProp = undefined as any;
        minDateProp = undefined as any;
        maxDateProp = undefined as any;

        currentFromScopeProps = undefined as any;
    });

    it(`should show closed / unfocused pickers by default`, () => {
        initializeWrapper();

        expect(currentFromScopeProps.props.focus).toBe(false);
        expect(currentToScopeProps.props.focus).toBe(false);
    });

    it(`should open dateFrom picker only when opening dateFrom picker`, async () => {
        initializeWrapper();

        currentFromScopeProps.handlers.change('2018-04-24');

        await Vue.nextTick();

        expect(currentFromScopeProps.props.focus).toBe(false);
        expect(currentToScopeProps.props.focus).toBe(true);
    });


    it(`should end selection when closing dateFrom picker without changing any value`, () => {
        initializeWrapper();

        expectSelectionEnd();
    });


    it(`should end selection when emptying dateFrom picker value`, () => {
        initializeWrapper();

        currentFromScopeProps.handlers.change(undefined);

        expectSelectionEnd();
    });

    it(`should end selection when emptying dateTo picker value`, () => {
        initializeWrapper();

        currentToScopeProps.handlers.change(undefined);

        expectSelectionEnd();
    });

    describe(`given no props are provided`, () => {
        it(`should pass down empty data to the first date`, () => {
            initializeWrapper();

            expect(currentFromScopeProps.props.value).toBe('');
            expect(currentFromScopeProps.props.min).toBe(undefined);
            expect(currentFromScopeProps.props.max).toBe(undefined);
        });

        it(`should pass down empty data to the second date`, () => {
            initializeWrapper();

            expect(currentToScopeProps.props.value).toBe('');
            expect(currentToScopeProps.props.min).toBe(undefined);
            expect(currentToScopeProps.props.max).toBe(undefined);
        });

        describe(`given dateFrom selector is open`, () => {
            beforeEach(() => {
                initializeWrapper();
            });

            it(`should end selection when emptying the value`, () => {
                currentFromScopeProps.handlers.change(undefined);

                expectSelectionEnd({ from: undefined, to: undefined });
            });

            it(`should display newly selected date when changing dateFrom value`, () => {
                const newFromDate: Date = new Date();
                currentFromScopeProps.handlers.change(newFromDate);

                expect(currentFromScopeProps.props.value).toEqual(MPeriodpicker.formatIsoDateToLocalString(newFromDate));
            });

            it(`should emit the new value when changing dateToValue`, async () => {
                const newFromDate: Date = new Date();
                const newToDate: Date = new Date();
                currentFromScopeProps.handlers.change(newFromDate);
                currentFromScopeProps.handlers.blur();
                currentToScopeProps.handlers.change(newToDate);
                currentToScopeProps.handlers.blur();
                await Vue.nextTick();
                expectSelectionEnd({ from: newFromDate, to: new ModulDate(newToDate).endOfDay() });
            });
        });

        describe(`given dateTo selector is open`, () => {
            beforeEach(() => {
                initializeWrapper();
            });

            it(`should end selection when emptying the value`, () => {
                currentToScopeProps.handlers.change(undefined);
                currentToScopeProps.handlers.blur();
                expectSelectionEnd({ from: undefined, to: undefined });
            });

            it(`should emit the new value when changing dateToValue`, async () => {
                const newToDate: Date = new Date();
                currentToScopeProps.handlers.change(newToDate);
                currentToScopeProps.handlers.blur();
                await Vue.nextTick();
                expectSelectionEnd({ from: undefined, to: new ModulDate(newToDate).endOfDay() });
            });
        });
    });

    describe(`given all props are provided`, () => {
        beforeEach(() => {
            fromDateProp = new Date();
            toDateProp = new Date(new Date().setDate(fromDateProp.getDate() + 1));
            minDateProp = new Date(new Date().setFullYear(fromDateProp.getFullYear() - 1));
            maxDateProp = new Date(new Date().setFullYear(fromDateProp.getFullYear() + 1));
        });

        it(`should pass down empty data to the first date`, () => {
            initializeWrapper();

            expect(currentFromScopeProps.props.value).toBe(MPeriodpicker.formatIsoDateToLocalString(fromDateProp));
            expect(currentFromScopeProps.props.min).toBe(minDateProp);
            expect(currentFromScopeProps.props.max).toBe(maxDateProp);
        });

        it(`should pass down empty data to the second date`, () => {
            initializeWrapper();

            expect(currentToScopeProps.props.value).toBe(MPeriodpicker.formatIsoDateToLocalString(toDateProp));
            expect(currentToScopeProps.props.min).toBe(fromDateProp);
            expect(currentToScopeProps.props.max).toBe(maxDateProp);
        });

        describe(`given dateFrom selector is open`, () => {
            beforeEach(() => {
                initializeWrapper();
            });

            it(`should end selection when emptying the value`, () => {
                currentFromScopeProps.handlers.change(undefined);

                expectSelectionEnd({ from: undefined, to: toDateProp });
            });

            it(`should not end selection and start dateTo pick when changing dateFrom the value`, () => {
                currentFromScopeProps.handlers.change(new Date());

                expectNoSelectionEnd();
                expect(wrapper.vm.fromIsFocused).toBe(false);
                expect(wrapper.vm.toIsFocused).toBe(true);
            });

            it(`should display newly selected date when changing dateFrom value`, () => {
                const newFromDate: Date = new Date();
                currentFromScopeProps.handlers.change(newFromDate);

                expect(currentFromScopeProps.props.value).toEqual(MPeriodpicker.formatIsoDateToLocalString(newFromDate));
            });

            it(`should emit the new value when changing dateToValue`, () => {
                const newFromDate: Date = new Date();
                const newToDate: Date = new Date();
                currentFromScopeProps.handlers.change(newFromDate);
                currentFromScopeProps.handlers.blur();
                currentToScopeProps.handlers.change(newToDate);
                currentToScopeProps.handlers.blur();
                expectSelectionEnd({ from: newFromDate, to: new ModulDate(newToDate).endOfDay() });
            });
        });

        describe(`given dateTo selector is open`, () => {
            beforeEach(() => {
                initializeWrapper();
            });

            it(`should end selection when emptying the value`, () => {
                currentToScopeProps.handlers.change(undefined);

                expectSelectionEnd({ from: fromDateProp, to: undefined });
            });

            it(`should emit the new value when changing dateToValue`, () => {
                const newToDate: Date = new Date();
                currentToScopeProps.handlers.change(newToDate);
                currentToScopeProps.handlers.blur();
                expectSelectionEnd({ from: fromDateProp, to: new ModulDate(newToDate).endOfDay() });
            });
        });
    });
});
