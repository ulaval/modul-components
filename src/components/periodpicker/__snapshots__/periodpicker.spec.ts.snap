// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`m-periodpicker should render correctly 1`] = `
<div>
    <m-validation-message-stub transition="true"></m-validation-message-stub>
</div>
`;
